### DISCUSSION TITLE
Copyright text is not recognized in SMF

### DISCUSSION THREAD

**Discussion by SinanKarasu** [orchetect/swift-midi#171] [Q&A] [answered]
As the title says.

**Comment by orchetect** [comment, marked as answer]
Copyright is an existing text SMF event type in MIDIKit. I checked the unit tests and it's functioning correctly.

What is the contents of `unrecognizedMeta`? What is the `metaType` and `data`? My guess is that it may not be an actual copyright text event that you're reading.

Copyright text is encoded as `0xFF 0x02 LEN <chars>`

Meta type should be `0x02` for copyright.

**Reply by SinanKarasu** [reply]
![Screenshot 2023-03-07 at 18 35 52](https://user-images.githubusercontent.com/7624146/223605563-4b1b929a-66f8-412c-9bb8-d03398d7a589.png)
If I download the following file
https://freemidi.org/download3-9779-it-doesnt-matter-alison-krauss
And parse it. I get the above screenshot. 
If I put the same file into MidiKit from the App Store, I get
![Screenshot 2023-03-07 at 18 37 58](https://user-images.githubusercontent.com/7624146/223605848-6f0d7d7e-5b9d-4c6f-addf-7c488473d612.png)
Notice that the spec says "This event should be the first event in the first track chunk, at time 0." but in this case, it is not. However MidiKit parses it anyway.

**Reply by SinanKarasu** [reply]
[MIDIKitSMF.zip](https://github.com/orchetect/MIDIKit/files/10916923/MIDIKitSMF.zip)
Here is what I'm trying to do, roughly. Very early stages. If you adopt it and make it an example, I'll copy it from you.

**Reply by orchetect** [reply]
Ah ok this is a regression in MIDIKit. I'll see if I can patch it soon.

**Reply by orchetect** [reply]
> "This event should be the first event in the first track chunk, at time 0." but in this case, it is not. 

Actually it is. All those events are at time 0. There's numerous event types that are supposed to be at time 0 but their precise order isn't important as long as that's the case.

**Reply by orchetect** [reply]
Ok bug is fixed and I just pushed release 0.8.6.

Side note: I noticed in MIDINoteNumber.swift that you have a raw table of constants. Not sure if you've discovered it yet, but MIDIKit has a type called [`MIDINote`](https://orchetect.github.io/MIDIKit/documentation/midikit/midinote) that can do pretty much all of that for you without using a table of constants.

**Reply by SinanKarasu** [reply]
rv.Second = MIDINote(event.note.number, style: .yamaha).name.stringValue()
			rv.Third = MIDINoteNumber.note(event.note.number).sharp
Note the first line gives me "C", the second line gives me "C4"
![Screenshot 2023-03-08 at 11 07 50](https://user-images.githubusercontent.com/7624146/223814259-4fe246c4-7544-479f-b69b-58df90f92ec0.png)

**Reply by SinanKarasu** [reply]
I'm going to adopt yours, and fix it later.

**Reply by orchetect** [reply]
```swift
MIDINote(event.note.number).name.stringValue() // "C"
```

This is returning base note name without the octave, so C is correct. I suppose the nomenclature could change and `name` could be called something less ambiguous like 'root' or 'degree'. `MIDINote` isn't really supposed to be a bastion of music theory, just a type for displaying MIDI note names according to typical DAW conventions.

```swift
MIDINote(event.note.number).stringValue() // "C4"
```

This what you want to return the full 'midi note name' string for display.

You can simplify this further, as `event.note` is already of type `MIDINote`:

```swift
event.note.stringValue() // "C4"
```

And if you haven't found it yet, you can get pretty-Unicode accidentals this way:

```swift
event.note.stringValue(unicodeAccidental: true)
```

**Reply by orchetect** [reply]
If you get any crashes or bugs that need addressing please open a new [Issue](https://github.com/orchetect/MIDIKit/issues) on GitHub for each one so I can track them and make them actionable.

Let's try to keep issues/discussion threads to one question or specific subject each so that they are easier to track and may better help others, instead of a single long thread containing multiple different concerns.

**Reply by orchetect** [reply]
`MIDIFile` and all its subtypes now gain Hashable conformance. As well, the MIDI file parsing crash is fixed. Will push a new release soon with those changes.

That said, there may be further issues with that MIDI file which caused it to attempt to read more data than it should have, but now that the crash is resolved the specific error should surface.
